### PR TITLE
Add BIP9 ActivationHeightProvider

### PR DESCRIPTION
**Pull request by quantumagi**
https://app.clickup.com/t/1uv60bv

### CODE REVIEW

File: src/Stratis.Bitcoin/Base/Deployments/ActivationHeightProvider.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        private ChainedHeader lastPollExpiryHeightChecked;
+        private int pollExpiryActivationHeight;
+
+        public ActivationHeightProvider(Network network, ThresholdConditionCache cache, ChainIndexer chainIndexer, int deployment)
+        {
+            this.network = network;
+            this.cache = cache;
+            this.chainIndexer = chainIndexer;
+            this.deployment = deployment;
+            this.lastPollExpiryHeightChecked = null;
```
Comment (zeptin) [inline src/Stratis.Bitcoin/Base/Deployments/ActivationHeightProvider.cs:25]: Are the names of these two variables related to a specific deployment this provider was intended to facilitate? It seems to me they could be called `lastHeightChecked` and `activationHeight` respectively (and the resulting clashing local `lastHeightChecked` renamed)

**Review by zeptin — COMMENTED** [review]
Looks good, just some nitpicks about variable names

**Comment by fassadlr** [comment]
@quantumagi this needs to point to 1.2.1

File: src/Stratis.Bitcoin/Base/Deployments/ActivationHeightProvider.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        public int ActivationHeight
+        {
+            get
+            {
+                if (this.activationHeight == int.MaxValue && this.chainIndexer.Tip != this.lastHeightChecked)
+                {
+                    if (this.lastHeightChecked != null)
+                        this.lastHeightChecked = this.chainIndexer.Tip.FindFork(this.lastHeightChecked);
+
+                    int lastHeightChecked = this.lastHeightChecked?.Height ?? 0;
```
Comment (zeptin) [inline src/Stratis.Bitcoin/Base/Deployments/ActivationHeightProvider.cs:42]: This should probably be renamed to prevent confusion

**Review by zeptin — APPROVED** [review]

**Pull request merged**